Feature asyncpg dialect doesn't support mutlihost connection string

iliadmitriev opened the pull request:
Fixes: #10004

This issues addresses the ability to create an engine with `asyncpg` postgresql dialect `postgresql+asyncpg`
with fallback alternative hosts.

* host1:5432 - main db for connections
* host2:5432 - fallback db

```python
import asyncio
from sqlalchemy import text
from sqlalchemy.ext.asyncio import create_async_engine

async def async_main() -> None:
    engine = create_async_engine(
        "postgresql+asyncpg://user:password@/dbname?host=host1:5432&host=host2:5432"
    )

    async with engine.begin() as conn:
        res = await conn.execute(
            text("select 1")
        )

    await engine.dispose()

asyncio.run(async_main())
```
How it works with `postgresql+psycopg2`
https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#specifying-multiple-fallback-hosts
Section: `Specifying multiple fallback hosts`

Also addresses
https://github.com/MagicStack/asyncpg/issues/352


### Description

+ added arguments parsing to `create_connect_args`
+ added tests `test_multihost_configuration`
+ added some docs with examples of configuration

### Checklist
<!-- go over following points. check them with an `x` if they do apply, (they turn into clickable checkboxes once the PR is submitted, so no need to do everything at once)

-->

This pull request is:

- [X] A documentation / typographical error fix
	- wrote some documentation with examples
- [X] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [x] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

iliadmitriev committed:
Feature asyncpg dialect doesn't support mutlihost connection string

+ moved base postgres dialect create arguments logic `PGDialect._split_multihost_from_url`
+ added asyncpg specific logic to `PGDialect_asyncpg.create_connect_args`
+ test case when one of hosts or ports is None
+ error message for case when one of hosts or ports is None
+ intelligeble error message when one of ports is not an integer

zzzeek commented:
do you think there's logic we can reuse here between asyncpg and psycopg?   id rather not have parsing duplicated   cc @CaselIT

iliadmitriev commented:
I got an idea, move all the logic to base class `PGDialect` from `PGDialect_asyncpg`.
Call superclass method from `PGDialect_asyncpg`
And then add options specific only for `asyncpg` (like `username` -> `user`)
I'll make it tomorrow.

iliadmitriev commented:
`PGDialect` doesn't have a `create_connect_args` method 
and has the child class `_PGDialect_common_psycopg(PGDialect)` which has `create_connect_args` method

`_PGDialect_common_psycopg` has two chldren
PGDialect_psycopg(_PGDialect_common_psycopg) has it's own `create_connect_args`, but refers to the superclass method first

PGDialect_psycopg2(_PGDialect_common_psycopg) hasn't method, and uses superclass `create_connect_args`

I think the best way to manage that is to move `_PGDialect_common_psycopg.create_connect_args` method to superclass `PGDialect`

iliadmitriev commented:
I made changes
Will you look @zzzeek ?

zzzeek commented:
I'm trying to edit this PR however I still dont know how to push to a github PR that's someone elses.   so i will push to gerrit instead

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 77fed83293d65f3303d95d1e981fe9a45306d372 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Failed to create a gerrit review, git squash against branch 'main' failed

zzzeek commented:
wow OK.   will make a new branch etc

zzzeek commented:
here's the git review : https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4755

CaselIT commented:
> I'm trying to edit this PR however I still dont know how to push to a github PR that's someone elses. so i will push to gerrit instead

the easiest way is to install the github cli https://cli.github.com/

Then it's quite easy:
- gh pr checkout 10005
- git commit ... # do not amend
- git push

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

thanks

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4755

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 63f023f145f79ab0d7afadb13cc408c9dd33dc59 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 63f023f145f79ab0d7afadb13cc408c9dd33dc59 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4755

zzzeek commented:
OK!   @iliadmitriev , take a look at the way I refactored this.   if you can confirm we did not lose the required functionality that would be helpful.

iliadmitriev commented:
Great Job! Thanks a lot!

The way how asyncpg library handles with connection argument `host`, lines 401,402:

https://github.com/MagicStack/asyncpg/blob/83c1159563b243faf30cfad9286153a63448ef6c/asyncpg/connect_utils.py#L387-L405

if passed argument host is not a list than it also puts it inside a list.

as soon as we pass a tuple of multiple hosts, it puts it inside a list of one element
and then it can't match the number of ports and then number of hosts and throws an exception.

<details>
  <summary>InterfaceError</summary>
  
  ```bash
  /Users/dmitriev/git/test_sa_asyncpg/venv/bin/python /Users/dmitriev/git/test_sa_asyncpg/main.py 
Traceback (most recent call last):
  File "/Users/dmitriev/git/test_sa_asyncpg/main.py", line 19, in <module>
    asyncio.run(async_main())
  File "/opt/homebrew/Cellar/python@3.11/3.11.4/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.4/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.4/Frameworks/Python.framework/Versions/3.11/lib/python3.11/asyncio/base_events.py", line 653, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/Users/dmitriev/git/test_sa_asyncpg/main.py", line 11, in async_main
    async with engine.begin() as conn:
  File "/opt/homebrew/Cellar/python@3.11/3.11.4/Frameworks/Python.framework/Versions/3.11/lib/python3.11/contextlib.py", line 204, in __aenter__
    return await anext(self.gen)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dmitriev/git/sqlalchemy/lib/sqlalchemy/ext/asyncio/engine.py", line 1047, in begin
    async with conn:
  File "/Users/dmitriev/git/sqlalchemy/lib/sqlalchemy/ext/asyncio/base.py", line 127, in __aenter__
    return await self.start(is_ctxmanager=True)  # type: ignore
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dmitriev/git/sqlalchemy/lib/sqlalchemy/ext/asyncio/engine.py", line 268, in start
    await greenlet_spawn(self.sync_engine.connect)
  File "/Users/dmitriev/git/sqlalchemy/lib/sqlalchemy/util/_concurrency_py3k.py", line 190, in greenlet_spawn
    result = context.throw(*sys.exc_info())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dmitriev/git/sqlalchemy/lib/sqlalchemy/engine/base.py", line 3264, in connect
    return self._connection_cls(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dmitriev/git/sqlalchemy/lib/sqlalchemy/engine/base.py", line 145, in __init__
    self._dbapi_connection = engine.raw_connection()
                             ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dmitriev/git/sqlalchemy/lib/sqlalchemy/engine/base.py", line 3288, in raw_connection
    return self.pool.connect()
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/dmitriev/git/sqlalchemy/lib/sqlalchemy/pool/base.py", line 452, in connect
    return _ConnectionFairy._checkout(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dmitriev/git/sqlalchemy/lib/sqlalchemy/pool/base.py", line 1267, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dmitriev/git/sqlalchemy/lib/sqlalchemy/pool/base.py", line 716, in checkout
    rec = pool._do_get()
          ^^^^^^^^^^^^^^
  File "/Users/dmitriev/git/sqlalchemy/lib/sqlalchemy/pool/impl.py", line 169, in _do_get
    with util.safe_reraise():
  File "/Users/dmitriev/git/sqlalchemy/lib/sqlalchemy/util/langhelpers.py", line 147, in __exit__
    raise exc_value.with_traceback(exc_tb)
  File "/Users/dmitriev/git/sqlalchemy/lib/sqlalchemy/pool/impl.py", line 167, in _do_get
    return self._create_connection()
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dmitriev/git/sqlalchemy/lib/sqlalchemy/pool/base.py", line 393, in _create_connection
    return _ConnectionRecord(self)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dmitriev/git/sqlalchemy/lib/sqlalchemy/pool/base.py", line 678, in __init__
    self.__connect()
  File "/Users/dmitriev/git/sqlalchemy/lib/sqlalchemy/pool/base.py", line 902, in __connect
    with util.safe_reraise():
  File "/Users/dmitriev/git/sqlalchemy/lib/sqlalchemy/util/langhelpers.py", line 147, in __exit__
    raise exc_value.with_traceback(exc_tb)
  File "/Users/dmitriev/git/sqlalchemy/lib/sqlalchemy/pool/base.py", line 898, in __connect
    self.dbapi_connection = connection = pool._invoke_creator(self)
                                         ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dmitriev/git/sqlalchemy/lib/sqlalchemy/engine/create.py", line 637, in connect
    return dialect.connect(*cargs, **cparams)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dmitriev/git/sqlalchemy/lib/sqlalchemy/engine/default.py", line 615, in connect
    return self.loaded_dbapi.connect(*cargs, **cparams)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dmitriev/git/sqlalchemy/lib/sqlalchemy/dialects/postgresql/asyncpg.py", line 919, in connect
    await_only(creator_fn(*arg, **kw)),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dmitriev/git/sqlalchemy/lib/sqlalchemy/util/_concurrency_py3k.py", line 125, in await_only
    return current.driver.switch(awaitable)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dmitriev/git/sqlalchemy/lib/sqlalchemy/util/_concurrency_py3k.py", line 185, in greenlet_spawn
    value = await result
            ^^^^^^^^^^^^
  File "/Users/dmitriev/git/test_sa_asyncpg/venv/lib/python3.11/site-packages/asyncpg/connection.py", line 2092, in connect
    return await connect_utils._connect(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dmitriev/git/test_sa_asyncpg/venv/lib/python3.11/site-packages/asyncpg/connect_utils.py", line 874, in _connect
    addrs, params, config = _parse_connect_arguments(timeout=timeout, **kwargs)
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dmitriev/git/test_sa_asyncpg/venv/lib/python3.11/site-packages/asyncpg/connect_utils.py", line 633, in _parse_connect_arguments
    addrs, params = _parse_connect_dsn_and_args(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dmitriev/git/test_sa_asyncpg/venv/lib/python3.11/site-packages/asyncpg/connect_utils.py", line 415, in _parse_connect_dsn_and_args
    port = _validate_port_spec(host, port)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/dmitriev/git/test_sa_asyncpg/venv/lib/python3.11/site-packages/asyncpg/connect_utils.py", line 166, in _validate_port_spec
    raise exceptions.InterfaceError(
asyncpg.exceptions._base.InterfaceError: could not match 2 port numbers to 1 hosts

Process finished with exit code 1
  ```
</details>


So, I suggest convert `host` and `port` from tuples to lists.

iliadmitriev commented on lib/sqlalchemy/dialects/postgresql/asyncpg.py:
```diff
@@ -1060,10 +1086,16 @@ def do_terminate(self, dbapi_connection) -> None:
         opts.update(url.query)
+
+        if multihosts:
+            opts["host"] = multihosts
+            opts["port"] = multiports
```
Here, I suggest convert `host` and `port` from tuples to lists.
asyncpg is waiting for a list.

iliadmitriev replied:
Also, asyncpg doesn't accept None values (And unfortunately don't use default values in case of None)
So, hosts and ports in a list should match, and shouldn't be None values

zzzeek replied:
yeah what should we do , raise an error?  I would rather not guess a default

zzzeek replied:
we would have this raise be on the asyncpg side; asyncpg gets the list of ports, does "if None in ports: raise ..."

iliadmitriev replied:
The way I see it, the best way to organize this is to conform asyncpg library as much as possible.
And as far as library don't want to get empty values, we should raise an exception.

Imagine situation some engineer didn't set configuration correctly (For example number of ports and host don't match). And if this exception is risen from asyncpg side, it will look something like this:
```
TypeError: int() argument must be a string, a bytes-like object or a real number, not 'NoneType'
```
Which is absolutely unreadable and very hard to understand what is going on, and where is it happened.

So, I suggest filter all the empty values from `multihosts` and `multiports`
, and check if their length match. (also we can check if each port is a string representing an integer)
Otherwise raise an `ArgumentError` exception with intelligible message.

iliadmitriev replied:
Something like this 

```python
if not all(multihosts):
    raise exc.ArgumentError(
        "Some hosts is not specified"
        " in a connection string"
    )
if not all(multiports) or not all(
        map(
            lambda x: isinstance(x, str) and x.isdigit(),
            multiports
        )
):
    raise exc.ArgumentError(
        "Some ports is not specified"
        " in a connection string"
    )
    
```

zzzeek replied:
OK give it a try.  add a new test for it in postgresql/test_dialect.py, using just asyncpg.dialect() the way the other tests do.  the tests for this feature shouldn't require that the underlying `asyncpg` library is actually imported (that is, dont use `create_engine()` in the test).

iliadmitriev replied:
@zzzeek done, have a look

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision c63a60396a9437dc37ec0c878bad9910f313de97 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset c63a60396a9437dc37ec0c878bad9910f313de97 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4755

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/asyncpg.py:
```diff
@@ -47,6 +47,32 @@
 
+.. _asyncpg_multihost_connecting:
+
+Multihost Connections
```
**mike bayer** (zzzeek) wrote:

have this be only a link to the existing psycopg section. formats are identical

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4755

iliadmitriev replied:
I made a commit, but I'm not sure this is it. Can you check or maybe show an example?

sqla-tester replied:
**mike bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4755

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/asyncpg.py:
```diff
@@ -1060,10 +1086,24 @@ def do_terminate(self, dbapi_connection) -> None:
         opts.update(url.query)
+
+        if multihosts:
+            if not all(multihosts):
```
**mike bayer** (zzzeek) wrote:

is there any way that None could be present in hosts here?   there should be a test if so

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4755

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/asyncpg.py:
```diff
@@ -1060,10 +1086,24 @@ def do_terminate(self, dbapi_connection) -> None:
+                    "Some hosts is not specified" " in a connection string"
+                )
+            if not all(multiports):
+                raise exc.ArgumentError(
```
**mike bayer** (zzzeek) wrote:

need test for this as well

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4755

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/asyncpg.py:
```diff
@@ -1060,10 +1086,24 @@ def do_terminate(self, dbapi_connection) -> None:
+                )
+            if not all(multiports):
+                raise exc.ArgumentError(
+                    "Some ports is not specified" " in a connection string"
```
**mike bayer** (zzzeek) wrote:

messages here should be:  "All hosts are required to be present for asyncpg multiple host URL"  "All ports are required to be present for asyncpg multiple host URL"

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4755

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3086,6 +3086,48 @@ def set_deferrable(self, connection, value):
+
+        if "port" in url.query:
+            if integrated_multihost:
+                raise exc.ArgumentError(
```
**mike bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4755

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3086,6 +3086,48 @@ def set_deferrable(self, connection, value):
+
+        if ports:
+            ports = tuple(int(x) if x else None for x in ports)
+
```
**mike bayer** (zzzeek) wrote:

we need a test that tests for ValueError if the port is not an int, and we might want to raise a nicer error

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4755

sqla-tester commented on test/dialect/postgresql/test_dialect.py:
```diff
@@ -1,8 +1,11 @@
 import logging
 import logging.handlers
 
+import pytest
```
**mike bayer** (zzzeek) wrote:

remove this...

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4755

sqla-tester commented on test/dialect/postgresql/test_dialect.py:
```diff
@@ -240,53 +244,138 @@ def test_psycopg2_nonempty_connection_string_w_query(self):
+    @testing.combinations(
+        (
+            "postgresql+asyncpg://USER:PASS@/DB?host=hostA",
+            pytest.raises(exc.ArgumentError),
```
**mike bayer** (zzzeek) wrote:

I would take all the "Raise" scenarios and put them as separate tests.  then we can't use pytest.raises() here, we have our own method "with expect_raises_message()" that has to be used (see line 391 for example)

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4755

sqla-tester commented on test/dialect/postgresql/test_dialect.py:
```diff
@@ -373,6 +462,29 @@ def test_connect_psycopg_multiple_hosts(self, pattern):
+        "host=H:P&host=H:P&host=H:P",
+        "host=H,H,H&port=P,P,P",
+    )
+    def test_connect_asyncpg_multiple_hosts(self, pattern):
```
**mike bayer** (zzzeek) wrote:

what does this test that the test_multi_asyncpg_hosts method does not test?  there should be just one set of tests for this if nothing is new here

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4755

iliadmitriev replied:
It tests a connection, not just parsing a dsn string arguments, like `test_multi_asyncpg_hosts`
I've copied it from `test_connect_psycopg_multiple_hosts` and adopted it for testing positive scenarios of asyncpg connection.

zzzeek replied:
OK why only asyncpg then?

iliadmitriev replied:
this is separate test for asyncpg only is because asyncpg have different positive scenarios (it doesn't set default port when port is None, and it throws exceptions)

there is the test called for testing connection of psycopg and psycopg2 `test_connect_psycopg_multiple_hosts`

sqla-tester reviewed (commented):
**mike bayer** (zzzeek) wrote:

nice, here's the next set of changes

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4755

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 7dace41852caf1419170030c4d1441f4378d6c8b of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Failed to create a gerrit review, git squash against branch 'main' failed

iliadmitriev commented:
> Failed to create a gerrit review, git squash against branch 'main' failed

@zzzeek what this message means? should I squash commits and rebase against top of main branch or just leave commits as they are and ignore this message?

zzzeek commented:
> > Failed to create a gerrit review, git squash against branch 'main' failed
> 
> @zzzeek what this message means? should I squash commits and rebase against top of main branch or just leave commits as they are and ignore this message?



> > Failed to create a gerrit review, git squash against branch 'main' failed
> 
> @zzzeek what this message means? should I squash commits and rebase against top of main branch or just leave commits as they are and ignore this message?

oops, sorry we had a big commit recently, yes please rebase against "main" again and we can pull it in.

This change is at the point where it just needs doc edits, mind if I just finish it up ?

iliadmitriev commented:
> > > Failed to create a gerrit review, git squash against branch 'main' failed
> > 
> > 
> > @zzzeek what this message means? should I squash commits and rebase against top of main branch or just leave commits as they are and ignore this message?
> 
> > > Failed to create a gerrit review, git squash against branch 'main' failed
> > 
> > 
> > @zzzeek what this message means? should I squash commits and rebase against top of main branch or just leave commits as they are and ignore this message?
> 
> oops, sorry we had a big commit recently, yes please rebase against "main" again and we can pull it in.
> 
> This change is at the point where it just needs doc edits, mind if I just finish it up ?

Done, I've rebased against upstream main branch and squashed all commits.

> This change is at the point where it just needs doc edits, mind if I just finish it up ?
Yeah, sure I would be very grateful.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 94bba62774377fd1654296c7ca4fe5114f75fcf5 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 94bba62774377fd1654296c7ca4fe5114f75fcf5 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4755

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3086,6 +3086,54 @@ def set_deferrable(self, connection, value):
+
+        if ports:
+            try:
+                ports = tuple(int(x) if x else None for x in ports)
```
**mike bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4755

sqla-tester commented on lib/sqlalchemy/dialects/postgresql/base.py:
```diff
@@ -3086,6 +3086,54 @@ def set_deferrable(self, connection, value):
+                ports = tuple(int(x) if x else None for x in ports)
+            except ValueError:
+                raise exc.ArgumentError(
+                    f"Some of specified ports is not a "
```
**mike bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4755

sqla-tester commented on test/dialect/postgresql/test_dialect.py:
```diff
@@ -240,53 +241,156 @@ def test_psycopg2_nonempty_connection_string_w_query(self):
+            "host=hostA:111&host=hostB:222&host=hostC:333",
+            {
+                "database": "DB",
+                "user": "USER",
```
**mike bayer** (zzzeek) wrote:

Done

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4755

sqla-tester reviewed (commented):
**mike bayer** (zzzeek) wrote:

this got very complicated as the multihost format has a *lot* of uncovered cases, and I tried to cover everything possible here, including urls that have a regular host plus tries to do multihost (error), urls that have multihost config but only one host or only one port, etc.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4755

* lib/sqlalchemy/dialects/postgresql/asyncpg.py (line 1094): Done
* lib/sqlalchemy/dialects/postgresql/asyncpg.py (line 1099): Done
* lib/sqlalchemy/dialects/postgresql/asyncpg.py (line 1100): Done
* test/dialect/postgresql/test_dialect.py (line 7): Done
* test/dialect/postgresql/test_dialect.py (line 470): Done

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

ok here goes...thanks very much!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4755

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4755 has been **merged**. Congratulations! :)
